### PR TITLE
Proxy letter avatars by default

### PR DESCRIPTION
**Pull request by mpalmer**
On sites that don't otherwise configure an avatar fallback, Discourse will
now tell the client to get its letter avatars from a location which nginx
proxies to the centralised `avatars.discourse.org` service.  This alleviates
privacy concerns, whilst still providing some degree of performance benefit
(no need for every site to delay avatar response by 300ms for image
rendering).

It is still possible to gain the benefits of global image caching and the
lower latency of requesting directly from a CDN, by explicitly changing the
`external_system_avatars_url` site setting to
`https://avatars.discourse.org/letter/{first_letter}/{color}/{size}.png`.

### CODE REVIEW

File: config/nginx.sample.conf [hunk 1]
```diff
@@ -192,6 +192,11 @@ server {
       break;
     }
 
+    location /letter_avatar_proxy {
+      rewrite /letter_avatar_proxy/(.*)$ /$1 break;
+      proxy_pass https://avatars.discourse.org;
+    }
```
Comment (calmh) [inline config/nginx.sample.conf]: By default I expect this to pass on all client sent headers, plus X-Forwarded-For and so on so I think it has negligible effect privacy wise.

File: app/models/user.rb [hunk 2]
```diff
@@ -475,6 +475,7 @@ def self.system_avatar_template(username)
       url.gsub! "{color}", letter_avatar_color(username.downcase)
       url.gsub! "{username}", username
       url.gsub! "{first_letter}", username[0].downcase
+      url.gsub! "{hostname}", RailsMultisite::ConnectionManagement.current_hostname
```
Comment (ZogStriP) [inline app/models/user.rb]: You should probably use `Discourse.current_hostname` here

File: config/site_settings.yml [hunk 3]
```diff
@@ -580,7 +580,7 @@ files:
     client: true
     shadowed_by_global: true
   external_system_avatars_url:
-    default: "https://avatars.discourse.org/letter/{first_letter}/{color}/{size}.png"
+    default: "//{hostname}/letter_avatar_proxy/letter/{first_letter}/{color}/{size}.png"
```
Comment (riking) [inline config/site_settings.yml]: Just use `/letter_avatar_proxy/letter/{first_letter}/{color}/{size}.png`, no need to include the hostname.

File: config/nginx.sample.conf (hunk 1 above)
Comment (riking) [inline config/nginx.sample.conf]: Add CF-Real-IP to the list.

File: app/models/user.rb (hunk 2 above)
Comment (mpalmer) [inline app/models/user.rb]: Tick.  I copied that more-or-less blindly from line 467.

File: config/site_settings.yml (hunk 3 above)
Comment (mpalmer) [inline config/site_settings.yml]: Tru 'dat.

File: config/nginx.sample.conf (hunk 1 above)
Comment (mpalmer) [inline config/nginx.sample.conf]: I'll go one better: `proxy_pass_request_headers off`...

File: config/site_settings.yml (hunk 3 above)
Comment (SamSaffron) [inline config/site_settings.yml]: I wonder what happens in a subdirectory setup though ...

Comment (mpalmer) [inline config/site_settings.yml]: That's a rather ugly thought... `/me adds to the test matrix`

Comment (konklone) [inline config/site_settings.yml]: It looks like the class methods to get the path prefix, if there is one, are available: https://github.com/discourse/discourse/blob/master/lib/discourse.rb#L184-L211

**Comment by mpalmer** [comment]
Another day, another PR update...

This version is unstoppable: it caches, it slices, it dices.  It doesn't leak request headers:

    2015/10/09 04:46:48 [debug] 256#0: *5 http request line: "GET /letter_avatar_proxy/letter/f/eada6e/120.png HTTP/1.1"
    2015/10/09 04:46:48 [debug] 256#0: *5 http uri: "/letter_avatar_proxy/letter/f/eada6e/120.png"
    2015/10/09 04:46:48 [debug] 256#0: *5 http args: ""
    2015/10/09 04:46:48 [debug] 256#0: *5 http exten: "png"
    2015/10/09 04:46:48 [debug] 256#0: *5 http header: "If-Modified-Since: Tue, 11 Jan 2000 00:57:26 GMT"
    2015/10/09 04:46:48 [debug] 256#0: *5 http header: "Host: localhost:8080"
    2015/10/09 04:46:48 [debug] 256#0: *5 http header: "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8"
    2015/10/09 04:46:48 [debug] 256#0: *5 http header: "User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.65 Safari/537.36"
    2015/10/09 04:46:48 [debug] 256#0: *5 http header: "DNT: 1"
    2015/10/09 04:46:48 [debug] 256#0: *5 http header: "Accept-Encoding: gzip, deflate, sdch"
    2015/10/09 04:46:48 [debug] 256#0: *5 http header: "Accept-Language: en-AU,en;q=0.8,en-US;q=0.6"
    2015/10/09 04:46:48 [debug] 256#0: *5 http header: "Cookie: <blah blah blah>"
    2015/10/09 04:46:48 [debug] 256#0: *5 http header: "Via: 1.1 mistress (squid/3.4.8)"
    2015/10/09 04:46:48 [debug] 256#0: *5 http header: "X-Forwarded-For: 192.168.121.1"
    2015/10/09 04:46:48 [debug] 256#0: *5 http header: "Cache-Control: max-age=0"
    2015/10/09 04:46:48 [debug] 256#0: *5 http header: "Connection: keep-alive"

    [...]

    2015/10/09 04:46:48 [debug] 256#0: *5 http script var: "avatars.discourse.org"
    2015/10/09 04:46:48 [debug] 256#0: *5 http proxy header:
    "GET /letter/f/eada6e/120.png HTTP/1.0
    Host: avatars.discourse.org
    Connection: close

    "

It doesn't send back response cookies (naughty Cloudflare!):

    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy status 200 "200 OK"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "Server: cloudflare-nginx"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "Date: Fri, 09 Oct 2015 04:46:45 GMT"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "Content-Type: image/png"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "Content-Length: 496"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "Connection: close"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "Set-Cookie: __cfduid=dccddce28254045f9cef929c5b28ec9bb1444366005; expires=Sat, 08-Oct-16 04:46:45 GMT; path=/; domain=.discourse.org; HttpOnly"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "Cache-Control: public, max-age=157788000"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "Last-Modified: Tue, 11 Jan 2000 00:57:26 GMT"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "Expires: Thu, 08 Oct 2020 10:46:45 GMT"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "Etag: F1202342181101"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "Vary: Accept-Encoding"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "CF-Cache-Status: HIT"
    2015/10/09 04:55:29 [debug] 903#0: *2 http proxy header: "CF-RAY: 2327718ed3860b26-SYD"

    [...]

    2015/10/09 04:55:29 [debug] 903#0: *2 HTTP/1.1 200 OK
    Server: nginx
    Date: Fri, 09 Oct 2015 04:55:29 GMT
    Content-Type: image/png
    Content-Length: 496
    Connection: keep-alive
    Cache-Control: public, max-age=157788000
    Last-Modified: Tue, 11 Jan 2000 00:57:26 GMT
    Expires: Thu, 08 Oct 2020 10:46:45 GMT
    Vary: Accept-Encoding
    CF-Cache-Status: HIT
    CF-RAY: 2327718ed3860b26-SYD

It also handles path-prefixed sites, which I was pleasantly surprised about.

So, I think I've addressed all the feedback.  I can haz merge plz?

**Comment by konklone** [comment]
:clap: :clap: :clap: This looks really solid.

**Comment by SamSaffron** [comment]
we better merge this :+1:

**Pull request merged**